Fix Android JIT compilation failure with template literals

anivar opened the pull request:
Fixes #1365

## Problem
Template literals cause interpretLoop method to exceed Android JIT compilation size limit (7392KB), resulting in 10x performance degradation.

## Solution
Extract template literal logic to proxy pattern, similar to existing RegExpProxy implementation.

## Changes
- Add TemplateLiteralProxy interface and implementation
- Delegate template literal operations from ScriptRuntime to proxy
- Reduce interpretLoop method size

## Testing
Local tests show performance restored:
- With fix: 0.016ms per operation
- Without fix: ~1000ms per operation (issue #1365)

Minimal fix that follows established Rhino patterns.

rPraml commented on rhino/src/main/java/org/mozilla/javascript/ScriptRuntime.java:
```diff
@@ -5828,6 +5828,21 @@ public static Scriptable wrapRegExp(Context cx, Scriptable scope, Object compile
             Context cx, Scriptable scope, Object[] strings, int index) {
+        TemplateLiteralProxy proxy = cx.getTemplateLiteralProxy();
+        if (proxy != null) {
+            return proxy.getCallSite(cx, scope, strings, index);
```
Can you explain to me what exactly is supposed to happen here? As far as I know, the same code is simply being called in a somewhat cumbersome way. The intention is that the JIT can then compile it more optimally?

rPraml reviewed (commented):
> * Removed ServiceLoader (was causing rhino-engine test issues)

Note: services have to be registered in META-INF (when you usen "old-school" classpath) and in module-info.java for module-path. (see RegexpProxy)

I have to admit that I only looked at the PR superficially and perhaps don't fully understand the problem on Android.

It would be great if we had a way to test this natively on Android in the future – preferably automatically – to ensure that the next person doesn't break it again.

Unfortunately, I don't know the best way to do this (Docker image with Android emulator?).
However, we don't have many people here with sufficient Android knowledge. Perhaps you do?

anivar commented:
> > * Removed ServiceLoader (was causing rhino-engine test issues)
> 
> Note: services have to be registered in META-INF (when you usen "old-school" classpath) and in module-info.java for module-path. (see RegexpProxy)

@rPraml You're right about ServiceLoader requiring META-INF/module-info - I initially used it following RegExpProxy, but it caused test failures so I simplified to direct instantiation.

  For Android testing, I agree automation would be valuable. I included manual test scripts in `/android-test/`. A Docker-based Android emulator in CI could work - happy to help set that up in a
  follow-up PR if there's interest.

rPraml commented:
Thanks, for your explanation, now I understand a bit better what's going on here. 

I’m wondering if this could be simplified a bit without having to touch so many classes (e. g. Context) .

My idea is, make TemplateLiteralProxy as `@FunctionalInterface` and put the `static TemplateLiteralProxy = ScriptRuntime::getTemplateLiteralCallsiteImpl` into the ScriptRuntime class. I think the value must not be final, kand maybe also public) so that the android compiler thinks, it could be changed and does not try to inline the code. 

If that’s sufficient, I’d see it as a more compact solution.
If not, and you’d like to merge your approach, please check the comments again — it still mentions "loading via ServiceLoader." 

One thing I’m not fully clear on: why is the getTemplateLiteralCallsiteImpl method called directly as a fallback when no service was found, and does that end up being inlined again? I have to admit I don’t have much Android experience here. 

> A Docker-based Android emulator in CI could work - happy to help set that up in a follow-up PR if there's interest.

I think this would help us a lot. 
(If adding this to CI is too much effort right now, having a script that can be run locally would already be a good improvement.)

anivar commented:
@rPraml Thank you for the excellent suggestion\! I've updated the PR with your simplified approach:

- Made `TemplateLiteralProxy` a simple functional interface (without @FunctionalInterface annotation for Android compatibility)
- Added a static non-final field in `ScriptRuntime` with method reference: `public static TemplateLiteralProxy templateLiteralProxy = ScriptRuntime::getTemplateLiteralCallSiteImpl`
- Removed `DefaultTemplateLiteralProxy` class entirely
- Removed proxy management code from `Context`
- Fixed the ServiceLoader comment issue

This achieves the same Android JIT fix with much cleaner code. The static non-final field prevents inlining just as effectively as the original implementation.

All tests pass with this simplified version. The Android performance improvement remains the same (~10x faster on Android devices).

The simplified implementation reduces the code complexity significantly while maintaining the fix's effectiveness.

Regarding the Android testing script - I've created a local Docker-based testing script in `android-test/test-android-local.sh` that can help verify the fix. As you suggested, this could be expanded to CI in a follow-up PR if there's interest.

anivar commented on tests/src/test/java/org/mozilla/javascript/tests/TemplateLiteralPerformanceTest.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @author Anivar Aravind
+ */
+public class TemplateLiteralPerformanceTest {
```
JMH cannot reproduce Android ART JIT behavior. The issue is specific to Android's runtime optimization at level -1. A desktop JMH benchmark would be testing the wrong runtime
  entirely.

SevenNight2012 commented:
> Fixes #1365
> 
> ## Problem
> Template literals cause interpretLoop method to exceed Android JIT compilation size limit (7392KB), resulting in 10x performance degradation.
> 
> ## Solution
> Extract template literal logic to proxy pattern, similar to existing RegExpProxy implementation.
> 
> ## Changes
> * Add TemplateLiteralProxy interface and implementation
> * Delegate template literal operations from ScriptRuntime to proxy
> * Reduce interpretLoop method size
> 
> ## Testing
> Local tests show performance restored:
> 
> * With fix: 0.016ms per operation
> * Without fix: ~1000ms per operation (issue [The performance of version 1.7.14 has gone backwards #1365](https://github.com/mozilla/rhino/issues/1365))
> 
> Minimal fix that follows established Rhino patterns.

Hello, i am the author of #1365, you mentioned "Android JIT compilation size limit (7392KB)", where can I check this parameter, or is there any documentation for reference?

anivar commented:
> > Fixes #1365
> > ## Problem
> > Template literals cause interpretLoop method to exceed Android JIT compilation size limit (7392KB), resulting in 10x performance degradation.
> > ## Solution
> > Extract template literal logic to proxy pattern, similar to existing RegExpProxy implementation.
> > ## Changes
> > 
> > * Add TemplateLiteralProxy interface and implementation
> > * Delegate template literal operations from ScriptRuntime to proxy
> > * Reduce interpretLoop method size
> > 
> > ## Testing
> > Local tests show performance restored:
> > 
> > * With fix: 0.016ms per operation
> > * Without fix: ~1000ms per operation (issue [The performance of version 1.7.14 has gone backwards #1365](https://github.com/mozilla/rhino/issues/1365))
> > 
> > Minimal fix that follows established Rhino patterns.
> 
> Hello, i am the author of #1365, you mentioned "Android JIT compilation size limit (7392KB)", where can I check this parameter, or is there any documentation for reference?

<img width="814" height="182" alt="Screenshot 2025-08-29 at 2 47 16 PM" src="https://github.com/user-attachments/assets/605e75f1-d36b-4f42-b316-913b23967c27" />
From this comment https://github.com/mozilla/rhino/issues/1365#issuecomment-1982336274
